### DISCUSSION TITLE
Random password generation method on Windows 10

### DISCUSSION THREAD

**Discussion by Mindean** [keepassxreboot/keepassxc#7014] [Q&A] [answered]
Hey guys, new with this software and have a question: 

I have read KeePassXC on Linux use the Linux kernel PRNG for random password generation, wich take into consideration things like keystrokes, mouse movements, date, time, OS, device information, among others.

I want to know if in the KeePassXC Windows 10 version there is an entropy pool too working for random password generation that takes into consideration this kind of things for randomness. 

Probably elaborate some? I have not found this information on the official web so would appreciate any information/resource.

Thanks!

**Comment by droidmonkey** [comment, marked as answer]
The current released version uses libgcrypt which pulls entropy from rndw32: https://www.gnupg.org/documentation/manuals/gcrypt/Random_002dNumber-Subsystem-Architecture.html#Random_002dNumber-Subsystem-Architecture

https://git.gnupg.org/cgi-bin/gitweb.cgi?p=libgcrypt.git;a=blob_plain;f=random/rndw32.c;hb=refs/heads/master

Version 2.7.0 and onward uses the Botan Cryptographic library which pulls entropy from RtlGenRandom (equivalent to CryptGenRandom): https://botan.randombit.net/handbook/api_ref/rng.html#entropy-sources

Read more on CryptGenRandom
https://docs.microsoft.com/en-us/windows/win32/api/wincrypt/nf-wincrypt-cryptgenrandom

At the end of the day, both libgcrypt and botan end up using the built in windows RNG.

**Comment by Mindean** [comment]
Thank you for such detailed answer, and all the work put on this great software!
